File: src/main/java/com/dv/telegram/command/SetSetting.java
package com.dv.telegram.command;

import com.dv.telegram.WikiBot;
import com.dv.telegram.config.BotSetting;

public class SetSetting extends BasicBotCommand {

    @Override
    public String getName() {
        return SetSetting.class.getSimpleName();
    }

    @Override
    public String getDescription(WikiBot bot) {
        return String.format(
            "`%s %s <settingName> <settingValue>` — установить значение настройки с названием `<settingName>` в значение `<settingValue>`.",
            bot.getBotName(),
            getCommandText()
        );
    }

    @Override
    public boolean useMarkdownInResponse() {
        return true;
    }

    @Override
    public String getDefaultCommandName() {
        return "/setSetting";
    }

    @Override
    public String getResponse(String text, WikiBot bot) {
        int commandStartIndex = text.indexOf(getCommandText());
        if (commandStartIndex < 0) {
            return unknownSettingResponse();
        }

        int commandEndIndex = commandStartIndex + getCommandText().length();
        if (commandEndIndex >= text.length()) {
            return unknownSettingResponse();
        }

        String settingNameAndValue = text.substring(commandEndIndex).trim();

        String nameValueSeparator = " ";
        int separatorIndex = settingNameAndValue.indexOf(nameValueSeparator);

        if (separatorIndex < 0 || separatorIndex >= settingNameAndValue.length()) {
            return unknownSettingResponse(); // todo: probably another more concrete response
        }

        String settingName = settingNameAndValue.substring(0, separatorIndex).trim();
        String settingValue = settingNameAndValue.substring(separatorIndex).trim();

        BotSetting<?> botSetting = bot.getSettings().getBotSetting(settingName);
        if (botSetting == null) {
            return unknownSettingResponse();
        }

        try {
            botSetting.setValue(settingValue);
            bot.getSettings().fillSettingCacheFields();
        }
        catch (Exception e) {
            return String.format("Ошибка при установке настройки *%s* в значение\n%s", settingName, BasicBotCommand.getSettingValueForMarkdown(settingValue));
        }

        return String.format("*%s* установлена в значение\n%s", botSetting.getName(), BasicBotCommand.getSettingValueForMarkdown(botSetting));
    }

    private String unknownSettingResponse() {
        return "Неизвестное имя настройки.";
    }
}

File: src/main/kotlin/com/dv/telegram/command/SetSetting.kt
package com.dv.telegram.command

import com.dv.telegram.WikiBot

class SetSetting : BasicBotCommand() {
    override val name: String = javaClass.simpleName

    override fun getDescription(bot: WikiBot) =
        "`${bot.botName} $commandText <settingName> <settingValue>` — установить значение настройки с названием `<settingName>` в значение `<settingValue>`."

    override fun useMarkdownInResponse() = true

    override val defaultCommandName = "/setSetting"

    override fun getResponse(text: String, bot: WikiBot): String {
        val commandStartIndex = text.indexOf(commandText)
        if (commandStartIndex < 0) {
            return unknownSettingResponse()
        }

        val commandEndIndex = commandStartIndex + commandText.length
        if (commandEndIndex >= text.length) {
            return unknownSettingResponse()
        }

        val settingNameAndValue = text.substring(commandEndIndex).trim()

        val nameValueSeparator = " "
        val separatorIndex = settingNameAndValue.indexOf(nameValueSeparator)

        if (separatorIndex < 0 || separatorIndex >= settingNameAndValue.length) {
            return unknownSettingResponse() // todo: probably another more concrete response
        }

        val settingName = settingNameAndValue.substring(0, separatorIndex).trim()
        val settingValue = settingNameAndValue.substring(separatorIndex).trim()

        val botSetting = bot.settings.getBotSetting(settingName)
            ?: return unknownSettingResponse()

        return try {
            botSetting.setValue(settingValue)
            bot.settings.fillSettingCacheFields()

            "*${botSetting.name}* установлена в значение\n${getSettingValueForMarkdown(settingValue)}"
        }
        catch (e: Exception) {
            "Ошибка при установке настройки *${botSetting.name}* в значение\n${getSettingValueForMarkdown(settingValue)}"
        }
    }

    private fun unknownSettingResponse(): String = "Неизвестное имя настройки."
}
